### PR TITLE
Ignore nonexisting submodule paths

### LINKED ISSUE #8366: [NBug] Crashing on startup when no submodule present
## Current behaviour

Git Extensions crashes on startup with some repos, but not others. Appears to be due to no submodules being present.

## Expected behaviour

Git Extensions starts up with all repos without crashing.

## Steps to reproduce

Open a repo with no submodules, it may crash.

## Error Details
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at GitUI.BranchTreePanel.RepoObjectsTree.SubmoduleTree.CreateSubmoduleNodes(IEnumerable`1 submodules, GitModule threadModule, List`1& nodes)
   at GitUI.BranchTreePanel.RepoObjectsTree.SubmoduleTree.FillSubmoduleTree(SubmoduleInfoResult result)
   at GitUI.BranchTreePanel.RepoObjectsTree.SubmoduleTree.<LoadNodesAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.BranchTreePanel.RepoObjectsTree.Tree.<ReloadNodesAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.BranchTreePanel.RepoObjectsTree.SubmoduleTree.<>c__DisplayClass4_0.<<OnStatusUpdated>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.ThreadHelper.<>c__DisplayClass13_0.<<FileAndForget>b__0>d.MoveNext()
```

## Additional information
Crashing caused by what I believe is no submodules being present. This crash doesn't occur with older versions of Git Extensions (2.48.05).

## Environment
- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4180.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by gerhardol**
Fixes #8366

## Proposed changes

Show MessageBox with submodule information

Replaces #9612, also some other situations handling #9056
This could be done in a nicer way, but this is important to point users in a better direction just not ignoring (as in 3.5) or showing a popup recommending to report an issue on GE (as in master).
For users, the behavior in master seem like a regression.

1. See #9612 - due to null types in master
Note that the popup is shown two times when opening a repo (structure and status) then periodically when status is updated. (So did the Validate error.)
The popup could be limited to be shown for just the first popup when the structure has been updated, but this is an error in the repo setup.

See this (auto closed) PR for how to recreate the issue.

2. Open when submodule path (not yet) exists due to #9056
  - Submodule Open in Sidepanel 
  - Submodule open in RevDiff/RevFileTree
  - Submodule open in Browse submodule toolbar

This is easiest done by deleting or renaming a submodule.
This also occurred when doubleclicking a submodule no longer existing, for instance the GitExtensionsDoc for commits two years ago.
Some occurrences like doubleclicking a deleted submodule in RevDiff was previously just ignored.

### Before

See also #9612 

image

### After

image

image

image

## Test methodology 

Manual

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
Should we provide a button/link to open submodules management dialog?

**Comment by gerhardol** [comment]
> Should we provide a button/link to open submodules management dialog?

I rather keep this simple, there are other PRs that I want to have merged...

**Review by mstv — COMMENTED** [review]
Use `MessageBoxes`, please.

**Comment by RussKie** [comment]
I'm worried that an average Joe user may not know what to do when these messages pop up. A link to "Manage submodules" will provide a significantly more positive experience.

**Comment by gerhardol** [comment]
> Use MessageBoxes, please.

Done

> I'm worried that an average Joe user may not know what to do when these messages pop up. A link to "Manage submodules" will provide a significantly more positive experience.

I just want something good enough. This is better than the igore or error popup in 3.5 or the popup after #9056 in master.
This can be improved sure, there are other issues to work on too.

File: GitUI/BranchTreePanel/RepoObjectsTree.Nodes.Submodules.cs [hunk 1]
```diff
@@ -93,6 +93,12 @@ protected override string NodeName()
 
             public void Open()
             {
+                if (!Directory.Exists(Info.Path))
+                {
+                    MessageBoxes.SubmoduleDirectoryDoesNotExist(null, Info.Path, Info.Text);
```
Comment (RussKie) [inline GitUI/BranchTreePanel/RepoObjectsTree.Nodes.Submodules.cs:98]: Give the null parameter name. Why is it null? The error will be unparented, and can get lost behind other windows.

Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.Nodes.Submodules.cs]: changed to `this` where possible (not in side panel or the generic Launch).

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 2]
```diff
@@ -2490,10 +2490,19 @@ private void RevisionInfo_CommandClicked(object sender, ResourceManager.CommandE
...
+            }
+
+            string path = menuSender.Tag as string;
+            if (!Directory.Exists(path))
+            {
+                MessageBoxes.SubmoduleDirectoryDoesNotExist(this, path);
+                return;
             }
+
+            SetWorkingDir(menuSender.Tag as string);
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormBrowse.cs:2505]:
```suggestion
            SetWorkingDir(path);
```

**Review by mstv — APPROVED** [review]
LGTM

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 2 above)
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormBrowse.cs:2505]: ok

**Comment by gerhardol** [comment]
Plan to squash merge tomorrow.
Possible enhancements can be added later.

**Pull request merged**